sdk/src/client/api/block_builder/transaction.rs

```diff
@@ -27,63 +28,85 @@ const SINGLE_UNLOCK_LENGTH: usize = 1 + 1 + Ed25519Signature::PUBLIC_KEY_LENGTH
 const REFERENCE_ACCOUNT_NFT_UNLOCK_LENGTH: usize = 1 + 2;
 
-/// Verifies the semantic of a prepared transaction.
-pub fn verify_semantic(
-    input_signing_data: &[InputSigningData],
-    transaction_payload: &SignedTransactionPayload,
-    mana_rewards: impl Into<Option<BTreeMap<OutputId, u64>>>,
-    protocol_parameters: ProtocolParameters,
-) -> Result<(), TransactionFailureReason> {
-    let inputs = input_signing_data
-        .iter()
-        .map(|input| (input.output_id(), &input.output))
-        .collect::<Vec<(&OutputId, &Output)>>();
+impl PreparedTransactionData {
+    /// Verifies the semantic of a prepared transaction.
+    pub fn verify_semantic(&self, protocol_parameters: &ProtocolParameters) -> Result<(), TransactionFailureReason> {
+        let inputs = self
+            .inputs_data
+            .iter()
+            .map(|input| (input.output_id(), &input.output))
+            .collect::<Vec<(&OutputId, &Output)>>();
+
+        let context = SemanticValidationContext::new(
+            &self.transaction,
+            &inputs,
+            None,
+            Some(&self.mana_rewards),
+            protocol_parameters,
+        );
+
+        context.validate()
+    }
+}
 
-    let context = SemanticValidationContext::new(
-        transaction_payload.transaction(),
-        &inputs,
-        Some(transaction_payload.unlocks()),
-        mana_rewards.into(),
-        protocol_parameters,
-    );
+impl SignedTransactionData {
+    /// Verifies the semantic of a prepared transaction.
+    pub fn verify_semantic(&self, protocol_parameters: &ProtocolParameters) -> Result<(), TransactionFailureReason> {
+        let inputs = self
+            .inputs_data
+            .iter()
+            .map(|input| (input.output_id(), &input.output))
+            .collect::<Vec<(&OutputId, &Output)>>();
 
-    context.validate()
+        let context = SemanticValidationContext::new(
+            &self.payload.transaction(),
+            &inputs,
+            Some(self.payload.unlocks()),
+            Some(&self.mana_rewards),
+            protocol_parameters,
+        );
+
+        context.validate()
+    }
 }
 
-/// Verifies that the signed transaction payload doesn't exceed the block size limit with 8 parents.
-pub fn validate_signed_transaction_payload_length(
-    signed_transaction_payload: &SignedTransactionPayload,
-) -> crate::client::error::Result<()> {
-    let signed_transaction_payload_bytes = signed_transaction_payload.pack_to_vec();
-    if signed_transaction_payload_bytes.len() > MAX_TX_LENGTH_FOR_BLOCK_WITH_8_PARENTS {
-        return Err(Error::InvalidSignedTransactionPayloadLength {
-            length: signed_transaction_payload_bytes.len(),
-            max_length: MAX_TX_LENGTH_FOR_BLOCK_WITH_8_PARENTS,
-        });
+impl SignedTransactionPayload {
+    /// Verifies that the signed transaction payload doesn't exceed the block size limit with 8 parents.
+    pub fn validate_length(&self) -> crate::client::error::Result<()> {
```

Comment:
Shouldn't we move this in the transaction builder finish()? Since we also the semantic validation there

Reply:
We don't sign there?

Reply:
I've created https://github.com/iotaledger/iota-sdk/issues/2119 btw, I'm not even sure we need these methods at all...

Reply:
Ah yeah but there is the other one on Transaction